libdnf/module/ModulePackageContainer.cpp

```diff
@@ -47,6 +47,12 @@ extern "C" {
 #include "modulemd/ModuleDefaultsContainer.hpp"
 #include "modulemd/ModuleProfile.hpp"
 
+
+template<>
+struct std::default_delete<DIR> {
```

Comment:
This way of adding to the `std` namespace seems to not work in gcc 4.8.2: https://github.com/rpm-software-management/libdnf/pull/745

However, why do you extend the `std` namespace with the `default_delete` specialization instead of using the `Deleter` template argument of `unique_ptr`? That is the way its supposed to be done, even though template specializations are allowed as an extension to `std`. Although is it even valid for the `DIR` type, which is from a 3rd party library (and strictly not program-defined?).

(see https://en.cppreference.com/w/cpp/language/extending_std)

Reply:
@j-mracek  I think that only simple change is needed to work with gcc 4.8.2:
```
namespave std {
template<>
struct default_delete<DIR> {
    void operator()(DIR * ptr) noexcept { closedir(ptr); }
};
}
```

Reply:
@lukash Extending std has some positives.
Source code is shorter and more readable.
The generated binary code is shorter, better optimised. Using of `Deleter ` costs some resources. Deleter (the pointer which value is not known during compile time) is stored to each unique_ptr instance during runtime.

Reply:
Readability is subjective, but I personally don't find it more readable. I find the use of `default_delete` specialization unexpected and if I am reading a declaration of `unique_ptr` without a deleter specified, I wouldn't expect there is a non-default one. The declaration of the deleter may be way up in the source file or even in a different file.

The optimisation aspect really seems marginal at best. Unless you're heavily optimising for performance, it doesn't really matter.

On the other hand:

- This solution seems more fragile. Imagine someone causing the compiler to not find the specialization by mistake, like due to a typo or moving the namespaces around, e.g. putting it into `libdnf::std` or global namespace etc. The compiler will silently fall back to the nonspecialized version of the template.

- The standard says this is only allowed for user-defined types, while DIR is from glibc, so I am not sure this is actually even allowed. Other parts of the program or even a library could potentially re-define the template specialization, breaking the one definition rule.

C++ standard clause 17.6.4.2.1, sentence 1:
> The behavior of a C++ program is undefined if it adds declarations or definitions to namespace std or to a namespace within namespace std unless otherwise specified. A program may add a template specialization for any standard library template to namespace std only if the declaration depends on a user-defined type and the specialization meets the standard library requirements for the original template and is not explicitly prohibited.

- C++ is hard enough to write and easy enough to make mistakes in. I strongly prefer to stick to the standard ways of doing things unless there is no other way. It is complex enough and error prone even without coming up with unusual and clever solutions that can easily backfire.

Reply:
@lukash DIR is user-defined type. But yes, glibc is not under our control. And there is some risk that one day glibc programmers will do change definition of DIR -> DIR will be a "alias" to native type (eg. int, terrible idea). And internally they will do reinterpret cast. And static type control for DIR type will be very weak.
OK we will discuss it personally.